MAINT: Fixes for NumPy deprecation and 3.10

larsoner opened the pull request:
1. `np.float` is just an alias for `float` so avoid the deprecation warning about it.
2. Change to `manylinux2014` for Python3.10 support
3. Add shims for VTK not having a 3.10 wheel (via `wheels.pyvistia.org`)

FYI it would be good to merge this and cut a release since there are no Python 3.10 wheels for `pymeshfix`.

larsoner commented:
I have a feeling `oldest-supported-numpy` might work for us here (somehow), but I'm just bumping the `requirements_build.txt` here hoping things come back green

larsoner commented on .github/workflows/testing-and-deployment.yml:
```diff
@@ -26,15 +36,31 @@ jobs:
-          pip install dist/*
 
-      - name: Install package
+      - name: Manually install VTK wheel on 3.10 since it does not work locally
+        run: |
+          pip install --upgrade pip
+          curl https://wheels.pyvista.org/vtk-9.1.0.dev0-cp310-cp310-macosx_12_0_x86_64.whl -o vtk-9.1.0.dev0-cp310-cp310-macosx_10_16_x86_64.whl
+          pip install ./vtk-9.1.0.dev0-cp310-cp310-macosx_10_16_x86_64.whl
+        if: ${{ matrix.os == 'macos-12' }}
```
Locally on my 12.3 Monterey install I cannot reproduce this issue, but on GH actions there is some problem installing this wheel despite many attempts at debugging. So here I just hack the filename to make it seem like at 10_16 wheel so that `pip` will install it.

It's a hack but I think it's acceptable since we want a 3.10 wheel for this package, and this hack can go away once VTK releases 9.2 with 3.10 wheels in the coming weeks (!?)

larsoner commented:
Finally green!

larsoner commented:
@pyvista/developers this one is ready for review. It would be nice to merge this so that we could get 3.10 wheels on PyPi

adeak commented on pymeshfix/cython/_meshfix.pyx:
```diff
@@ -197,12 +197,12 @@ cdef class PyTMesh:
-                v = np.ascontiguousarray(v, dtype=np.float)
+            if v.dtype not in (float, np.float64):
+                v = np.ascontiguousarray(v, dtype=np.float64)
             else:
                 v = np.ascontiguousarray(v)
-        elif v.dtype != np.float:
-            v = v.astype(np.float)
+        elif v.dtype not in (float, np.float64):
```
`v.dtype == float` can never be true, right? Either it's a numpy floating type or `object`.

In fact this snippet always converts to contiguous double, right?
```py
>>> arr = np.arange(3.)

>>> np.ascontiguousarray(arr, dtype=np.float_) is arr
True
```

Unless I'm missing some edge case this _whole_ if could be replaced with this one call. If the array is already C-contiguous double, it will be left intact. Otherwise it will be copied to a contiguous double array.

larsoner commented:
... I went a step further and just called `ascontiguousarray` with a `dtype`, as it should essentially be a no-op / just create a view if it's already contiguous, e.g.:
```
>>> import numpy as np
>>> x = np.array([1.], np.float64)
>>> y = np.ascontiguousarray(x, dtype=np.float64)
>>> y[0] = 0
>>> x
array([0.])
```
I also fixed some plain `assert` statements to raise a proper error, and changed to bare `raise Exception`s to `raise RuntimeError`s

adeak commented:
> ... I went a step further and just called ascontiguousarray with a dtype, as it should essentially be a no-op / just create a view if it's already contiguous, e.g.:

Yeah, that's what I had in mind.

> I also fixed some plain `assert` statements to raise a proper error, and changed to bare `raise Exceptions` to `raise RuntimeErrors`

Good idea, thanks. I wonder if we should define some more meaningful errors for these... not sure it's worth it for purity's sake.

larsoner commented:
All green again, okay to merge @adeak ?

adeak commented:
Sorry, @larsoner, I can't review any of the CI changes :( The numpy part seems good to me.

larsoner commented:
I think the CI stuff is necessary to get 3.10 support. If anyone else wants to look and finds problems, I'm happy to open another PR, or revert and try again! But in the meantime I'll go ahead and merge and cut a release to see if PyPi wheels show up.
